src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,439 @@
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if (!is_curve_supported(curve_str)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    /*
+     * Note that we use a buffer suitable for a public key to store a private
+     * key. This was a historical mistake, now preserved for compatibility.
+     */
+    output_length = ec_pub_key_size(curve_str, compressed);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    *key_length = output_length;
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_PRIV_KEY /* "priv" */, &d)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if ((output_length - sizeof(soter_container_hdr_t))
+        != bn_encode(d, (unsigned char*)(key + 1), output_length - sizeof(soter_container_hdr_t))) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    memcpy(key->tag, ec_priv_key_tag(curve_str), SOTER_CONTAINER_TAG_LENGTH);
+    key->size = htobe32(output_length);
+    soter_update_container_checksum(key);
+    *key_length = output_length;
+    res = SOTER_SUCCESS;
+
+err:
+
+    return res;
+}
+
+soter_status_t soter_ec_pub_key_to_engine_specific(const soter_container_hdr_t* key,
+                                                   size_t key_length,
+                                                   soter_engine_specific_ec_key_t** engine_key)
+{
+    const char* curve_str;
+    OSSL_PARAM params[3] = {[2] = OSSL_PARAM_END};
+    EVP_PKEY_CTX* ctx = NULL;
+    soter_status_t res;
+
+    if ((!key) || (key_length < sizeof(soter_container_hdr_t))) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (key_length != be32toh(key->size)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    /* Validate tag */
+    if (memcmp(key->tag, EC_PUB_KEY_PREF, strlen(EC_PUB_KEY_PREF)) != 0) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (SOTER_SUCCESS != soter_verify_container_checksum(key)) {
+        return SOTER_DATA_CORRUPT;
+    }
+
+    switch (key->tag[3]) {
+    case EC_SIZE_TAG_256:
+        curve_str = SN_X9_62_prime256v1;
+        break;
+    case EC_SIZE_TAG_384:
+        curve_str = SN_secp384r1;
+        break;
+    case EC_SIZE_TAG_521:
+        curve_str = SN_secp521r1;
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME /* "group" */,
+                                                 (char*)curve_str,
+                                                 strlen(curve_str));
+
+    params[1] = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_PUB_KEY /* "pub" */,
+                                                  (unsigned char*)(key + 1),
+                                                  (int)(key_length - sizeof(soter_container_hdr_t)));
+
+    ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
```

Comment:
Is there some constant for the magic `"EC"` string as well?

Reply:
Nope, everywhere I see usage of `EVP_PKEY_CTX_new_from_name()` in OpenSSL sources, it's string literal with values like `EC`, `DH`, `RSA`, `X25519`, `HKDF` and so on. Lowercase variants were also observed, so case is probably not important. Still, uppercase looks cooler